### ISSUE TITLE
gpt2 and cache activations

### ISSUE THREAD

**Issue by BlastWind** [BlastWind/hech-interp#1]
Support the activation cache as seen in [ARENA](https://arena-ch1-transformers.streamlit.app/[1.2]_Intro_to_Mech_Interp).

**Comment by BlastWind** [comment]
More precisely, we want a typed version of the rather concise `transformer_lens`'s [`run_with_cache`](https://transformerlensorg.github.io/TransformerLens/generated/code/transformer_lens.HookedTransformer.html#transformer_lens.HookedTransformer.HookedTransformer.run_with_cache). Probably some sort of type-level map is needed to keep track of the dimensions of different addressable layers —

![Image](https://github.com/user-attachments/assets/5e5de8f7-6689-4358-b29b-ca83487119aa)

**Comment by BlastWind** [comment]
Thoughts on implementation:

Goal: We provide the model's specification during cache-enabled inference to get a dependent cache map. We would know the exact shapes and names of the activation caches. For example, we can query `cache_map["attn"][7]["q"]` to get the attention query vector that was computed in the forward pass, for the 7th transformer block. With multiheaded attention, the query becomes more complicated.

Have: Currently, `infer` returns the logits (which is something to cache), but no additional cache objects. The cache should contain activations, not parameters. For example, `cache_map["attn"][7]["W_q"]` wouldn't make sense, the cache has no notion of the `q` weights. That's something that the concrete model should know of.

Gap: Who's gonna fill in the `cache_map`? And how are different keys going to be filled in?

Ideas: 
- Lift `infer` into a `State` monad where the state is the cache map. 
- The caller of the functions such as `forward` and `multiheadAttention` collect the subroutines' results and bin them into the right index within the cache. For the subroutines themselves do not know their own location and can't perform the proper cache updating.
- Can there be a more declarative way that directly associates cache indexes and keys with function calls, allowing for minimal code
- Do we need to construct a blank cache map when starting `infer` (all of the cache components are initialized to have 0s in the tensors)? Or can we build the full cache map up component-by-component during inferencing, such that, by the end of the inference, we also get a guarantee that the cache is constructed correctly?

**Comment by BlastWind** [comment]
Continuing the last brainstorms —

First of all, I'm going to ignore making things DRY. I'm just going to write a `infer_with_cache` function that copies the contents of `infer`. We might be able to do some sort of wrapping to reuse `infer` and its subroutines within `infer_with_cache`, but I'm gonna focus on the cache collection itself first. 

### Cache Collection and its Totality Guarantee

I think I really want to make the following guarantee: Inference with cache will certainly fulfill a cache map whose expected keys are explicitly specified.

We currently use a dependent map to model the cache map. While a dependent map sounds fancy, its main use is to enable keys to map to values of different type. Totality is not an implicit guarantee, just like how not all regular map are total (take a `Map String Int`, we don't need to define a pairing for every `String` in existence). 

I would like to make it total. My reasoning follows: There are finitely many components, and their outputs must all be cached, so our construct here is a total map. Hence, why not make it formally total? This will allow us to know that the cache we return back 

Rambling on this subject — I don't want any "fake totality" that is given for free by turning a `Map k v` into a `TotalMap k (Maybe v)`. While this map is total, the runtime check is simply delayed from at-the-key to at-the-(maybe)-value.

So somehow, we want to be able to incrementally construct this total map (since each subroutine should fill a subset of the cache map) and don't use any hacks. This feels paradoxical. How can increment construction (which suggests partial calculations) lead to totality?

In the future, I would also like to extend `infer_with_cache` to take in a cache filter. I.e., the caller can specify which components it wishes to cache. If a component is ignored, it won't show up in the type of the returned cache map.

Implementation plan:
1. Complete the `ActivationCache` dependent map type.
2. Write an `infer_with_cache` function. This probably warrants another files, for there are quite some subroutines to edit to seep the caching logic in.
3. At this point, querying the map returned from `infer_with_cache` works. However, the map lookup returns a `Maybe v`, even though we know that our map is total. We could use the unsafe lookup and leave a comment saying that it will be fine. But maybe we can do better — explore totality
4. Explore selective indexing

**Comment by BlastWind** [comment]
With https://github.com/BlastWind/hech-interp/pull/9, this is done. Later refactorings about the caching data structures might happen, see #7, #8 , #10. But activations are being cached.
